### DISCUSSION TITLE
Getting started, and Gnome Builder, LSP and swift-format

### DISCUSSION THREAD

**Discussion by olifink** [AparokshaUI/Adwaita#30] [General]
First of all @david-swift congrats on AparokshaUI, it's an amazing piece of work and very well documented. I  even found your tutorial pathway explained the "SwiftUI" concepts better than Apples official documentation. Well done :clap: 

I'm using it on Fedora/40 at the moment and here are a couple of things I found out that others perhaps might find helpful as well. I had the same issues with auto completion as described in #9 and decided to dig deeper as I also didn't quite understand the `flatpak` setup in builder. 

With the setup described in the tutorial I didn't get _proper_ autocompletion, only parts were working but crucially `import Adwaita` was failing:

![Screenshot from 2024-05-16 09-58-39](https://github.com/AparokshaUI/adwaita-swift/assets/323250/014bb832-31dd-4a39-87ff-c74b7351c15b)

After a lot of trial and error and fiddling with project settings, I could narrow it down to the fact that `sourcekit-lsp` relies not only on a properly built project for their indices, but also that `swift build` doesn't seem to provide them (fully?) when building with `release` configuration as specified in the build config in [io.github.AparokshaUI.AdwaitaTemplate.json](https://github.com/AparokshaUI/AdwaitaTemplate/blob/fd3b6d2bdd8db4074e02706cc33991e4036862ed/io.github.AparokshaUI.AdwaitaTemplate.json#L43) 

Although `-c release` makes perfect sense to distribute the project, changing it to `-c debug` during development now also let LSP complete the missing `Adwaita` package:

![Screenshot from 2024-05-16 10-17-52](https://github.com/AparokshaUI/adwaita-swift/assets/323250/f30a9e06-c592-449e-a869-c16f87fd1c1a)

Another option I found is to open a `New Build Terminal` (ctrl+alt+shift+t) and navigate to the source directory and do a manual `swift build` there. One note that you also might night to reopen the project in builder to make it recognize the index. I also tried `--enable-index-store` as a build option, but that didn't seem to help.

Now with this fixed I still had the remaining issue of the source formatter not working. This should be taken care of by `swift-format` however all I got was:

![Screenshot from 2024-05-16 09-55-55](https://github.com/AparokshaUI/adwaita-swift/assets/323250/9266595d-f54c-4a22-8e13-c97058cde660)

I'm afraid the I only workaround I found for this was with a local install of `swift-format` on my host system. I assume it might have something to do with the LSP and the formatter perhaps running in a different container environment and not having access to the same index files. I initially had a local build from the [swift-format repo](https://github.com/apple/swift-format) in conjunction with the fedora `swift-lang` rpm. However the rpm is still at swift 5.8 so in the end I found it much easier to just use [homebrew](https://brew.sh/) on linux (shudder!) to have  both swift 5.10 and swift-format on the host. 

WIth that setup, even the formatter now work for me in Builder. 

https://github.com/AparokshaUI/adwaita-swift/assets/323250/58610559-ab3d-4c8f-979e-928b9a8fd07b

The only issue left for me is that hover documentation windows in Builders are extremely fickle and there is no keyboard shortcut yet to open a more permanent view, but that's a problem for all other project types as well, not only Aparoksha...

**Comment by david-swift** [comment]
Thank you so much! This is the solution!

I think I'll change the template repository to build for debugging and add instructions [here](https://github.com/AparokshaUI/AdwaitaTemplate) advising to switch to release for publishing.

I don't know about the `swift-format` problem, maybe @JCWasmx86 does.

**Reply by JCWasmx86** [reply]
Regarding the formatter problem: This should only occur, if the swift code is invalid. Maybe it has something to do with version mismatches, e.g. the swift-format version shipped doesn't support a grammar yet, but it's supported in Swift already, I really don't know and I didn't do Swift for several months so I can't help debug this further

**Comment by olifink** [comment]
you're right @JCWasmx86 - I've checked the version of `swift-format` in the build container and it reports `508.0.0` while swift is at `5.10-RELEASE` so I think just updating it in `org.freedesktop.Sdk.Extension.swift5` should fix it.

![image](https://github.com/AparokshaUI/adwaita-swift/assets/323250/605b8f21-e088-4065-8975-12f14dfa02d0)

So I suspect the culprit is [swift-format-setup-offline.sh](https://github.com/flathub/org.freedesktop.Sdk.Extension.swift5/blob/branch/23.08/swift-format-setup-offline.sh)

**Reply by david-swift** [reply]
Thanks for reporting this, I'll update the SDK extension!

**Comment by olifink** [comment]
great! 

also fyi - I was just looking into the Builder sources to see what it would take to support Swift as a project template:

It seems they have two template plugins at the moment, one is  `meson-templates` the other is `make-templates`. Since SPM is neither I think it would not be too difficult to rework the smaller make plugin into a new `swiftpm-template` plugin and have it generate the content from [AdwaitTemplate](https://github.com/AparokshaUI/AdwaitaTemplate) - that should make it much easier to get started without the search/replace currently needed.

so unless somebody is already looking into it, I will give it a try. My C is a little rusty but I should manage, maybe @JCWasmx86 you could review or help. I see you've already contributed into the builder codebase, which would be a first time for me.

**Reply by JCWasmx86** [reply]
I didn't contribute to Builder in the last few months, but I would be happy to help. You can DM me either on Matrix (Link in Profile) or Discord

**Reply by david-swift** [reply]
That would be very cool! Thanks in advance, @olifink and @JCWasmx86!

**Comment by david-swift** [comment]
Okay, so I fixed the auto completion problem in https://github.com/AparokshaUI/AdwaitaTemplate/commit/d046bbca1399bf8bfbce721722b97362f91eb7bb.

The swift format problem should be fixed with https://github.com/flathub/org.freedesktop.Sdk.Extension.swift5/compare/branch/23.08...swift-lint-510-1-0 (`swift-format -v` outputs `510.1.0`).

One question to @JCWasmx86 now that you're here: is there a reason you built SwiftLint from source instead of using the Linux download (https://github.com/realm/SwiftLint/releases/download/0.55.1/swiftlint_linux.zip)?

**Reply by JCWasmx86** [reply]
> One question to @JCWasmx86 now that you're here: is there a reason you built SwiftLint from source instead of using the Linux download (https://github.com/realm/SwiftLint/releases/download/0.55.1/swiftlint_linux.zip)?

I didn't see that binary

**Reply by david-swift** [reply]
@olifink the swift-format problem should be fixed in the most recent SDK Extension version on Flathub. Could you confirm this?

**Reply by olifink** [reply]
@david-swift well, yes and no :face_with_diagonal_mouth: the `swift-format` is now the correct version in the platform SDK which is good, and it formats the file correctly when used directly in the build terminal, but I still get the same error banner trying to format from within the the ide.

Since I have Builder checked-out now anyway for the template work, I figured I could debug into the swift-format plugin. I probably need some time wrap my head about what's going... but what has been confusing for me so far is that the ide obviously picks up the `swift-format` binary on my host when it's there and fails when it's not there. 

Debugging through `create_run_context` in  `plugins/swift-format/gbp-swiftformat-formatter.c` seems to confirm that: it tries to find the `swift-format` via `ide_runtime_contains_program_in_path` and failing that, falls through to the `run_on_host:` label. 

So, I'm guessing @JCWasmx86  that `ide_config_get_runtime` is actually only going to return the *runtime* container (that I would also get via _New Runtime Terminal_ which doesn't have the `swift-format` installed) and not the *build* container with the sdk-extension for swift5...Maybe?

**Reply by olifink** [reply]
for reference:

```c
  if (ide_context_has_project (context))
    {
      IdeConfigManager *config_manager = ide_config_manager_from_context (context);
      IdeConfig *config = ide_config_manager_get_current (config_manager);
      IdeRuntime *runtime;

      if ((runtime = ide_config_get_runtime (config)) &&
          ide_runtime_contains_program_in_path (runtime, argv0, NULL))
        {
          ide_runtime_prepare_to_build (runtime, NULL, run_context);
          ide_run_context_append_argv (run_context, argv0);

          return g_steal_pointer (&run_context);
        }
    }

run_on_host:
  ide_run_context_push_host (run_context);
  ide_run_context_append_argv (run_context, argv0);
```

**Reply by olifink** [reply]
yup, stepping through it confirms that `ide_runtime_contains_program_in_path()` returns 0 and then it tries to run something on host that doesn't exist (unless you install it explicitly as I did to get it working)

So I think we need to find a way to run it via the build container context somehow

**Reply by olifink** [reply]
hmm, reading through Hergerts Builder Developer Notebook, it sounds like `ide_runtime_prepare_to_build` should do that. So this must be called before `ide_runtime_contains_program_in_path`.

But even if I change that, the sdk's `swift-format` is still not found... :confused:

**Reply by olifink** [reply]
quick update @david-swift  and @JCWasmx86 - after spelunking a while through various Builder matrix and irc channels I finally raised it as an [issue in gnome gitlab](https://gitlab.gnome.org/GNOME/gnome-builder/-/issues/2215) and maintainer Christian Hergert already pushed the fix. 

I just checked it out and can confirm the `swift-format` issue is now fixed and works directly without the need of a local install. 

So, I will now focus on a swift project template.
